lib/haskell/natural4/src/LS/Types.hs

```diff
@@ -214,43 +217,53 @@ type HornClause2 = HornClause BoolStructR
 
 -- | Though this is still quite 'syntactic'; would be even better if it were repns for a specific semantics
 data SimpleHlike a b c =
-  MkSimpleHL { shcGiven :: a
-             , shcRet :: b
-             , baseHL :: BaseHL
-             , shcSrcRef :: SrcRef
-             -- ^ may want to parametrize this
-             , shRLabel :: c
+  MkSimpleHL { shClauseGiven :: a
+             -- ^ The 'GIVEN' part of a horn-like rule.
+             , shClauseReturnVars :: b
+             -- ^ The 'GIVETH' part of a horn-like rule.
+             , shRuleSrcRef :: SrcRef
+             -- ^ Source reference to where the rule was defined.
+             -- While there is 'SrcRef' is usually generated by the
+             -- csv file parser, there is nothing that ties it inherently into
+             -- the csv pipeline.
+             --
+             -- TODO: Perhaps we want to be able to parameterise the type of
+             -- source reference we might have. For example, to have Source Spans
+             -- instead of only Source References.
+             , shBaseClauses :: BaseHL
+             -- ^ Clauses of the horn-like rule. Contains at least one clause.
+             , shRuleLabel :: c
```

Comment:
I definitely went overboard with the abbreviations in the first version of this; glad to see that you are using clearer names.

Bit of a nitpick, but: I find the `sh` prefix in `shRuleSrcref` a bit confusing given that you also use the `shClause` prefix --- I wonder if it might be better to re-use the convention from above and go for something like `shClauseRuleSrcref`. And similarly with `shBaseClauses` and `shRuleLabel`

Reply:
Hehe, yeah, the fewer non-standard abbreviation, the fewer names I have to remember :)

Good suggestion! During the renaming I was thinking, the givens and giveths could be specific to a single `AtomicHC`. But they can't! The `SimpleHlike` describes a full `Hornlike` Rule of the form `GIVEN ... GIVETH .. DECIDE ...`, right? The prefix could then be just `sh`, or at most `shRule`? `shClause` feels definitely wrong now, as I would associate only the body of the `DECIDE` part as "clauses". I think
```haskell
  MkSimpleHL { shGiven :: a
             -- ^ The 'GIVEN' part of a horn-like rule.
             , shReturnVariables :: b
             -- ^ The 'GIVETH' part of a horn-like rule.
             , shSrcRef :: SrcRef
             -- ^ Source reference to where the rule was defined.
             -- While there is 'SrcRef' is usually generated by the
             -- csv file parser, there is nothing that ties it inherently into
             -- the csv pipeline.
             --
             -- TODO: Perhaps we want to be able to parameterise the type of
             -- source reference we might have. For example, to have Source Spans
             -- instead of only Source References.
             , shClauses :: BaseHL
             -- ^ Clauses of the horn-like rule. Contains at least one clause.
             , shLabel :: c
             }
```
is pretty good. What do you think?

Reply:
This is reminding me of / making me think about bigger, more fundamental questions like, 
* if ultimately what we want is a functional language, then do we even want to talk about 'HornClauses' (which is more of a logic programming concept)? What would be the point of doing so? (Though, yes, I know there's still some uncertainty over whether what is wanted is a functional or logic programming language.)
* Are 'AtomicHC' and 'Hornlike' really appropriate names? Is it too counterintuitive that the givens and giveths are scoped, not to specific AtomicHCs, but rather to a group of them?

Not saying that you need to resolve these questions now --- more just things to maybe have at the back of your mind.

The more specific thoughts I have on this are:
* This seems fine overall, yeah, though now I'm starting to wonder if `shReturnVariables` is too specific, given that we're parameterizing over the type of that.
* Relatedly, does it even make sense for all of `a`, `b`, and `c`, to be type variables? (That was almost definitely my fault --- I reach for type variables too quickly.) It's hard to tell what a `SimpleHL` 'really is', when there's that many type variables.

Also, sidenote: at least as of last year, the SrcRefs that the parser supplies aren't always accurate. I don't know if the bug(s) have been fixed, but I would definitely do some testing before relying on them for anything important (or use the better parser(s) that you and Andres have in the works :)

Reply:
I think there is a lot of value of calling it a horn like rule since it is a well known concept. As long as we are still working on the Rules (e.g. traverse, transpile, etc...) I think the name is well-chosen. In the functional backend, I think we shouldn't have the concept of "Hornlike".